### PR TITLE
Fix build with latest flatbuffers

### PR DESCRIPTION
**Pull request by mattn**
This change make be possible to build gpu delegate on Linux.
For raspberry pi, this may not work with good performance since  this patch use libegl1-mesa, libgles2-mesa not VideoCore. This change does not provide changes for bazel BUILD since this is not for vc. However, if someone implement EGL3 for raspberry pi, this will work well.

Update https://github.com/tensorflow/tensorflow/issues/27320

### CODE REVIEW

**Comment by mattn** [comment]
To compiling, [Makefile](https://gist.github.com/ddc0fc9dc53f8a02f5f4fa953a29fe2d) is required.

File: tensorflow/lite/delegates/gpu/gl/egl_context.h [hunk 1]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 #define TENSORFLOW_LITE_DELEGATES_GPU_GL_EGL_CONTEXT_H_
 
 #include <vector>
+#include <cstring>
```
Comment (mattn) [inline tensorflow/lite/delegates/gpu/gl/egl_context.h]: This should move to egl_context.cc

**Review by shahzadlone — DISMISSED** [review]

File: tensorflow/lite/delegates/gpu/gl/kernels/elementwise.cc [hunk 2]
```diff
@@ -16,6 +16,7 @@ limitations under the License.
 #include "tensorflow/lite/delegates/gpu/gl/kernels/elementwise.h"
 
 #include <string>
+#include <iostream>
```
Comment (impjdi) [inline tensorflow/lite/delegates/gpu/gl/kernels/elementwise.cc]: we don't want to depend on iostream.

please remove this line and remove the std::cerr below.  that's bad leftover from someone's debugging.

File: tensorflow/lite/delegates/gpu/gl/egl_context.cc [hunk 3]
```diff
@@ -15,6 +15,9 @@ limitations under the License.
 
 #include "tensorflow/lite/delegates/gpu/gl/egl_context.h"
 
+#include <vector>
+#include <cstring>
```
Comment (impjdi) [inline tensorflow/lite/delegates/gpu/gl/egl_context.cc]: please sort alphabetically

```
#include <cstring>
#include <vector>
```

**Review by impjdi — CHANGES_REQUESTED** [review]

**Review by impjdi — CHANGES_REQUESTED** [review]
1. Changes to egl_context.cc are not required, right?  I don't see any cstring or vector usages.
2. I cannot approve #undef Status.  That would break Google.  I suggest that you exclude X11 locally when you build your stuff.
3. Changes to egl_context.h, elementwise.cc, Makefile, download_dependencies.sh look good to me.

**Comment by gbaned** [comment]
@mattn  Did you get a chance to look on reviewer comments? Please let us know on the update. Thanks!

**Comment by mattn** [comment]
> Changes to egl_context.cc are not required, right? I don't see any cstring or vector usages.

Okay, I'll update this PR in later.

> I cannot approve #undef Status. That would break Google. I suggest that you exclude X11 locally when you build your stuff.

Any change to add hack something to undefine macro?

> Changes to egl_context.h, elementwise.cc, Makefile, download_dependencies.sh look good to me.

Thank you.

**Comment by mattn** [comment]
@impjdi For about changing of egl_context.cc, calling `strstr` should be `std::strstr` ?

**Review by impjdi — CHANGES_REQUESTED** [review]
All changes except #undef Status look fine.  Could you please remove them, and keep those for yourself?

**Comment by hoonkai** [comment]
@mattn Hi I was trying to run your `Makefile`. Can I ask how you managed to get `compiled_model_generated.h` generated? I believe `serialization.h` needs it.

**Comment by mattn** [comment]
@impjdi removed changes of undef Status.

**Review by impjdi — DISMISSED** [review]

**Comment by gbaned** [comment]
@mattn Can you please check build failures? Thanks!

**Comment by gbaned** [comment]
@mattn Could you please check build failures and resolve the conflicts? Thanks!

**Comment by mattn** [comment]
Sorry, at least, I can't build this on raspberry pi since I removed "#define Status".  If you point the URL to the CI, I can fix it.

**Comment by gbaned** [comment]
@mattn Could you please resolve the conflicts? Thanks!

**Comment by mattn** [comment]
Sorry, `$(OBJDIR)%.o: %.cc %.cpp` did not work since the source of a target must be separated.

**Review by impjdi — DISMISSED** [review]

**Comment by mattn** [comment]
FYI, This is [SSD app using go-tflite](https://github.com/mattn/go-tflite/tree/master/_example/ssd). 7 May 2019

https://twitter.com/mattn_jp/status/1125429350495473664

And this is same app with this patch (including optimizations merged into master branch recently) 25 Jun 2019

https://twitter.com/mattn_jp/statuses/1143533459375214592

File: tensorflow/lite/tools/make/download_dependencies.sh [hunk 4]
```diff
@@ -37,6 +37,7 @@ NEON_2_SSE_URL="https://github.com/intel/ARM_NEON_2_x86_SSE/archive/master.zip"
 FARMHASH_URL="http://mirror.tensorflow.org/github.com/google/farmhash/archive/816a4ae622e964763ca0862d9dbd19324a1eaf45.tar.gz"
 FLATBUFFERS_URL="http://mirror.tensorflow.org/github.com/google/flatbuffers/archive/v1.11.0.tar.gz"
 FFT2D_URL="http://mirror.tensorflow.org/www.kurims.kyoto-u.ac.jp/~ooura/fft2d.tgz"
+FP16_URL="https://github.com/Maratyszcza/FP16/archive/master.zip"
```
Comment (miaout17) [inline tensorflow/lite/tools/make/download_dependencies.sh]: Could you point to a specific commit ID or tag?

**Review by miaout17 — CHANGES_REQUESTED** [review]
Could you explain more about how this works?

It seems the Makefile doesn't compile `delegates/egl/*` by default. 
Do you have a local Makefile change to make it compile with the newly included `FP16` library?

Thanks!

File: tensorflow/lite/tools/make/download_dependencies.sh (hunk 4 above)
Comment (mattn) [reply tensorflow/lite/tools/make/download_dependencies.sh]: AFAIK, fp16 provide only header file. I'll fix the `master.zip` to specific commit ID.

Comment (mattn) [reply tensorflow/lite/tools/make/download_dependencies.sh]: BTW, intel/ARM_NEON_2_x86_SSE is used with master.zip

https://github.com/tensorflow/tensorflow/blob/4247d8e498d047dcf9a718638a931d68591d8f91/tensorflow/lite/tools/make/download_dependencies.sh#L36

**Comment by impjdi** [comment]
@mattn

I ran into this problem of Status conflicting with X11 int when trying to use TFLite GPU on desktop.  It turns out EGL/egl.h pulls in EGL/eglplatform.h which pulls in X11/Xlib.h & X11/Xutil.h.  To bypass that, I built things with with

-DMESA_EGL_NO_X11_HEADERS

Of course, this may be different on RPi, but I would suggest looking tat the EGL include headers and make things work to exclude X11 includes.

**Comment by mattn** [comment]
>-DMESA_EGL_NO_X11_HEADERS

Oh, Thanks. I didn't know it.

**Comment by mattn** [comment]
You don't like to add `-DDMESA_EGL_NO_X11_HEADERS` in Makefile. Right?

**Comment by impjdi** [comment]
@mattn 

> You don't like to add `-DMESA_EGL_NO_X11_HEADERS` in Makefile. Right?

Hm, I am not a direct owner of TFLite (I'm only an owner of the GPU backend) and I don't know the answer to that :(  AFAIK I think the TF team is planning on deprecating the Makefile.  I'll loop in @jdduke but he's out for the week.  You might get an answer to your question next week.

**Comment by mattn** [comment]
I don't have strong opinion for deprecating Makefile, but deprecating Makefile may lose many tensorflow lite novices, or hobby users.

File: tensorflow/lite/tools/make/download_dependencies.sh (hunk 4 above)
Comment (miaout17) [reply tensorflow/lite/tools/make/download_dependencies.sh]: I'm sorry if I missed something obvious and asking naive question... 

It seems the current Makefile doesn't compile `delegates/egl/*` at all. 
So how does this change affect anything? Do you need to make a local change on `Makefile` to include `egl/` directory?

Comment (mattn) [reply tensorflow/lite/tools/make/download_dependencies.sh]: Yes, but tensorflow team have plan to deprecate Makefile.

Comment (jdduke) [reply tensorflow/lite/tools/make/download_dependencies.sh]: @mattn I'm still not quite following the rationale for this CL. You say that it fixes the gpu/ build for the Makefile, but you're not actually compiling anything in gpu/. If you can verify a Makefile build for the rpi 4, we would be willing to consider such a pull request. However, without actually pulling in gpu/ to the Makefile, I don't think it's worth pulling in the fp16 library (unless the current standard TFLite Makefile build is broken, without the GPU code).

Comment (mattn) [reply tensorflow/lite/tools/make/download_dependencies.sh]: If you allow me to put new Makefile on gpu directory, I'll add it. But you have plan to deprecate Makefiles.

**Comment by gbaned** [comment]
@mattn Could you please resolve the conflicts? Thanks!

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by mattn** [comment]
Well, I can't use the button.

**Pull request closed without merging**